File: mvpcore/src/main/java/biz/dealnote/mvp/core/AbsPresenter.java
package biz.dealnote.mvp.core;

import android.os.Bundle;
import android.support.annotation.CallSuper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.lang.ref.WeakReference;
import java.lang.reflect.InvocationTargetException;
import java.util.Set;
import java.util.concurrent.atomic.AtomicInteger;

import biz.dealnote.mvp.reflect.AnnotatedHandlerFinder;
import biz.dealnote.mvp.reflect.EventHandler;

/**
 * Created by admin on 11.07.2016.
 * mvpcore
 */
public abstract class AbsPresenter<V extends IMvpView> implements IPresenter<V> {

    /**
     * V (View Host - это не значит, что существуют реальные вьюхи,
     * с которыми можно работать и менять их состояние. Это значит,
     * что существует, к примеру, фрагмент, но нет гарантии, кто метод onCreateView был выполнен.
     * К примеру, фрагмент находится в стэке, он существует, может хранить какие-то данные,
     * но при перевороте его вьюв был уничтожен, но не был заново создан, ибо фрагмент не в топе
     * контейнера и не added)
     */
    private WeakReference<V> viewReference;
    private boolean guiCreated;

    private static final AtomicInteger IDGEN = new AtomicInteger();
    private int mPresenterId;

    public AbsPresenter(@Nullable Bundle savedInstanceState){
        if(savedInstanceState != null){
            mPresenterId = savedInstanceState.getInt(SAVE_ID);
            if(mPresenterId >= IDGEN.get()){
                IDGEN.set(mPresenterId + 1);
            }
        } else {
            mPresenterId = IDGEN.incrementAndGet();
        }
    }

    private static final String SAVE_ID = "save_presenter_id";

    public int getPresenterId(){
        return mPresenterId;
    }

    @CallSuper
    protected void onViewHostAttached(@NonNull V view) {

    }

    @CallSuper
    protected void onViewHostDetached() {

    }

    @CallSuper
    protected void onGuiCreated(@NonNull V viewHost) {

    }

    private void executeAllResolveViewMethods(){
        Set<EventHandler> resolveMethodHandlers =
                AnnotatedHandlerFinder.findAllOnGuiCreatedHandlers(this, AbsPresenter.class);

        for(EventHandler handler : resolveMethodHandlers){
            if(!handler.isValid()) continue;

            try {
                handler.handle();
            } catch (InvocationTargetException e) {
                e.printStackTrace();
            }
        }
    }

    @CallSuper
    protected void onGuiDestroyed() {

    }

    private boolean destroyed;

    private boolean guiResumed;

    public boolean isDestroyed() {
        return destroyed;
    }

    public boolean isGuiResumed() {
        return guiResumed;
    }

    @CallSuper
    protected void onGuiResumed() {

    }

    @CallSuper
    protected void onGuiPaused() {

    }

    @Override
    public final void destroy() {
        destroyed = true;
        onDestroyed();
    }

    @Override
    public final void resumeView() {
        guiResumed = true;
        onGuiResumed();
    }

    public boolean isViewHostAttached(){
        return viewReference.get() != null;
    }

    @Override
    public final void pauseView() {
        guiResumed = false;
        onGuiPaused();
    }

    @Override
    public final void attachViewHost(@NonNull V view) {
        viewReference = new WeakReference<>(view);
        onViewHostAttached(view);
    }

    @Override
    public final void detachViewHost() {
        viewReference = new WeakReference<>(null);
        onViewHostDetached();
    }

    @Override
    public final void createView(@NonNull V view) {
        guiCreated = true;
        executeAllResolveViewMethods();
        onGuiCreated(view);
    }

    @Override
    public final void destroyView() {
        guiCreated = false;
        onGuiDestroyed();
    }

    @CallSuper
    public void onDestroyed() {

    }

    @Override
    @CallSuper
    public void saveState(@NonNull Bundle outState) {
        outState.putInt(SAVE_ID, mPresenterId);
    }

    public final boolean isGuiReady() {
        return guiCreated;
    }

    //@Nullable
    public V getView() {
        return viewReference == null ? null : viewReference.get();
    }

    protected void callView(ViewAction<V> action){
        if(isGuiReady()){
            action.call(getView());
        }
    }
}

File: mvpcore/src/main/java/biz/dealnote/mvp/core/AbsPresenter.kt
package biz.dealnote.mvp.core

import android.os.Bundle
import android.support.annotation.CallSuper
import biz.dealnote.mvp.reflect.AnnotatedHandlerFinder
import java.lang.ref.WeakReference
import java.lang.reflect.InvocationTargetException
import java.util.concurrent.atomic.AtomicInteger

/**
 * Created by admin on 11.07.2016.
 * mvpcore
 */
abstract class AbsPresenter<V : IMvpView>(savedInstanceState: Bundle?) : IPresenter<V> {

    /**
     * V (View Host - это не значит, что существуют реальные вьюхи,
     * с которыми можно работать и менять их состояние. Это значит,
     * что существует, к примеру, фрагмент, но нет гарантии, кто метод onCreateView был выполнен.
     * К примеру, фрагмент находится в стэке, он существует, может хранить какие-то данные,
     * но при перевороте его вьюв был уничтожен, но не был заново создан, ибо фрагмент не в топе
     * контейнера и не added)
     */
    private var viewReference: WeakReference<V> = WeakReference<V>(null)

    var isGuiReady: Boolean = false

    var id: Int

    var isDestroyed: Boolean = false

    var isGuiResumed: Boolean = false

    val isViewHostAttached: Boolean
        get() = viewReference.get() != null

    val view: V?
        get() = viewReference.get()

    init {
        if (savedInstanceState != null) {
            id = savedInstanceState.getInt(SAVE_ID)
            if (id >= IDGEN.get()) {
                IDGEN.set(id + 1)
            }
        } else {
            id = IDGEN.incrementAndGet()
        }
    }

    @CallSuper
    protected open fun onViewHostAttached(view: V) {

    }

    @CallSuper
    protected fun onViewHostDetached() {

    }

    @CallSuper
    protected open fun onGuiCreated(viewHost: V) {

    }

    private fun executeAllResolveViewMethods() {
        val resolveMethodHandlers = AnnotatedHandlerFinder.findAllOnGuiCreatedHandlers(this, AbsPresenter::class.java)

        for (handler in resolveMethodHandlers) {
            try {
                handler.handle()
            } catch (ignored: InvocationTargetException) {

            }
        }
    }

    @CallSuper
    protected open fun onGuiDestroyed() {

    }

    @CallSuper
    protected open fun onGuiResumed() {

    }

    @CallSuper
    protected open fun onGuiPaused() {

    }

    override fun destroy() {
        isDestroyed = true
        onDestroyed()
    }

    override fun resumeView() {
        isGuiResumed = true
        onGuiResumed()
    }

    override fun pauseView() {
        isGuiResumed = false
        onGuiPaused()
    }

    override fun attachViewHost(view: V) {
        viewReference = WeakReference(view)
        onViewHostAttached(view)
    }

    override fun detachViewHost() {
        viewReference = WeakReference<V>(null)
        onViewHostDetached()
    }

    final override fun createView(view: V) {
        isGuiReady = true
        executeAllResolveViewMethods()
        onGuiCreated(view)
    }

    override fun destroyView() {
        isGuiReady = false
        onGuiDestroyed()
    }

    @CallSuper
    open fun onDestroyed() {

    }

    @CallSuper
    override fun saveState(outState: Bundle) {
        outState.putInt(SAVE_ID, id)
    }

    protected fun callView(action: ViewAction<V>) {
        if (isGuiReady) {
            view?.run {
                action.call(this)
            }
        }
    }

    companion object {
        private val IDGEN = AtomicInteger()
        private const val SAVE_ID = "save_presenter_id"
    }
}